core/src/HIP/Kokkos_HIP_Instance.cpp

```diff
@@ -0,0 +1,462 @@
+
+    Record::increment(r);
+
+    m_scratchSpace = reinterpret_cast<size_type *>(r->data());
+  }
+
+  return m_scratchSpace;
+}
+
+Kokkos::Experimental::HIP::size_type *HIPInternal::scratch_flags(
+    const Kokkos::Experimental::HIP::size_type size) {
+  if (verify_is_initialized("scratch_flags") &&
+      m_scratchFlagsCount * sizeScratchGrain < size) {
+    m_scratchFlagsCount = (size + sizeScratchGrain - 1) / sizeScratchGrain;
+
+    typedef Kokkos::Impl::SharedAllocationRecord<Kokkos::Experimental::HIPSpace,
+                                                 void>
+        Record;
+
+    Record *const r = Record::allocate(
+        Kokkos::Experimental::HIPSpace(), "InternalScratchFlags",
+        (sizeScratchGrain * m_scratchFlagsCount));
+
+    Record::increment(r);
+
+    m_scratchFlags = reinterpret_cast<size_type *>(r->data());
+
+    hipMemset(m_scratchFlags, 0, m_scratchFlagsCount * sizeScratchGrain);
+  }
+
+  return m_scratchFlags;
+}
+
+//----------------------------------------------------------------------------
+
+void HIPInternal::finalize() {
+  HIP().fence();
+  was_finalized = 1;
+  if (0 != m_scratchSpace || 0 != m_scratchFlags) {
+    typedef Kokkos::Impl::SharedAllocationRecord<Kokkos::Experimental::HIPSpace>
+        RecordHIP;
+
+    RecordHIP::decrement(RecordHIP::get_record(m_scratchFlags));
+    RecordHIP::decrement(RecordHIP::get_record(m_scratchSpace));
+
+    m_hipDev            = -1;
+    m_hipArch           = -1;
+    m_multiProcCount    = 0;
+    m_maxWarpCount      = 0;
+    m_maxBlock          = 0;
+    m_maxSharedWords    = 0;
+    m_maxShmemPerBlock  = 0;
+    m_scratchSpaceCount = 0;
+    m_scratchFlagsCount = 0;
+    m_scratchSpace      = 0;
+    m_scratchFlags      = 0;
+  }
+}
+
+//----------------------------------------------------------------------------
+
+Kokkos::Experimental::HIP::size_type hip_internal_cu_count() {
+  return HIPInternal::singleton().m_multiProcCount;
+}
+
+Kokkos::Experimental::HIP::size_type hip_internal_maximum_extent_size() {
+  return HIPInternal::singleton().m_maxWarpCount;
+}
+
+Kokkos::Experimental::HIP::size_type hip_internal_maximum_shared_words() {
+  return HIPInternal::singleton().m_maxSharedWords;
+}
+
+Kokkos::Experimental::HIP::size_type *hip_internal_scratch_space(
+    const Kokkos::Experimental::HIP::size_type size) {
+  return HIPInternal::singleton().scratch_space(size);
+}
+
+Kokkos::Experimental::HIP::size_type *hip_internal_scratch_flags(
+    const Kokkos::Experimental::HIP::size_type size) {
+  return HIPInternal::singleton().scratch_flags(size);
+}
+
+}  // namespace Impl
+}  // namespace Experimental
+}  // namespace Kokkos
+
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+namespace Experimental {
+
+// HIP::size_type HIP::detect_device_count()
+//{ return Impl::HIPInternalDevices::singleton().m_hipDevCount ; }
+
+int HIP::concurrency() {
+#if defined(KOKKOS_ARCH_KAVERI)
```

Comment:
Lets change this to something useful. I.e. at least some GPU which is actually supported ...

Reply:
Sure but I don't know how you came up with these numbers

Reply:
figure out how many threads can run on the VEGA GPUs. You can also try and actually ask it i.e. what we do in CUDA.